src/containers/methodCallContainer.js

```diff
@@ -1,55 +1,134 @@
-import React, { useContext, useCallback } from 'react';
+import React, { useContext, useCallback, useState, useEffect } from 'react';
 import { NeedMethodMessage, NeedURLMessage, MethodCall } from '../components';
 import { AppContext, LogContext } from '../context';
 import Web3RpcCalls from '../helpers/web3Config';
 import buildProvider from '../helpers/buildProvider';
+import {
+  fetchOrParseAbi,
+  getFilteredMethods,
+  getContractFriendlyArguments,
+  getFormInputsFromMethod,
+  onUpdateAbi,
+} from '../helpers/contracts';
 import { navigate, useParams } from '@reach/router';
 
+const CONTRACT_FUNCTION_METHOD = 'eth_call';
+
 const MethodCallContainer = () => {
   const params = useParams();
-  const { codeSampleVisible, toggleSampleCode } = useContext(AppContext);
+  const { codeSampleVisible, toggleSampleCode, abi, setAbi } = useContext(
+    AppContext
+  );
   const { addToLog } = useContext(LogContext);
   const logItem = useCallback(addToLog, []);
+
   const {
     web3URL = '',
     web3Lib = '',
     currentMethod = '',
     formArgs = '',
   } = params;
-  const argumentList = formArgs.split('/');
+
   const web3Method = Web3RpcCalls[currentMethod] || {};
   const { description, disabled } = web3Method || {};
-  const { args, exec } = web3Method[web3Lib] || {};
-  const setArgumentList = (val, index) => {
-    const argsCopy = [...argumentList];
-    argsCopy[index] = val;
-    // bad pattern?
-    let joinedArgs = argsCopy.join('/');
+  const { args: initialFormInputs, exec } = web3Method[web3Lib] || {};
+
+  const [formInputs, setFormInputs] = useState([]);
+  const [argumentList, setArgumentList] = useState([]);
+
+  const updateURL = (val, index) => {
+    let argsList = formArgs.split('/').slice(0, formInputs.length); // Remove dangling arguments
+    argsList[index] = val;
+    let joinedArgs = argsList.join('/');
     let url = `/${web3URL}/${web3Lib}/`;
     if (currentMethod) url += `${currentMethod}/`;
-    if (args.length > 0) url += `${joinedArgs}`;
+    if (formInputs.length > 0) url += `${joinedArgs}`;
     navigate(url);
   };
-  const runRequest = (args) => {
+
+  const onUpdateArguments = async (val, index) => {
+    if (currentMethod === CONTRACT_FUNCTION_METHOD) {
```

Comment:
can we combine this line and the one after?

```javascript
if (currentMethod === CONTRACT_FUNCTION_METHOD && index === 1) {
```

Reply:
Done.